### PR TITLE
feat/gas-refund-improvements: PSP-IP20 - 7d virtual lockup (part 2/2)

### PR DESCRIPTION
**Pull request by mwamedacen**
https://vote.paraswap.network/#/proposal/0xf5f492264f6bdeb8b65b55ee30d6b9e86191a6999154416bd677e65a40148f8a

See diff between without vs with 7d lockup: https://www.diffchecker.com/sCcBiSFH

### CODE REVIEW

File: scripts/gas-refund-program/staking/common.ts [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+  timestamp: number,
+  stakeAtStartOfVirtualLockup: BigNumber,
+  differentialStates: TimeSeries | undefined,
+) {
+  const startOfVirtualLockupPeriod = timestamp - VIRTUAL_LOCKUP_PERIOD;
+
+  if (!differentialStates || differentialStates.length === 0)
+    return stakeAtStartOfVirtualLockup;
+
+  const minStakeHoldDuringVirtualLockup = (differentialStates || []).reduce(
```
Comment (alexshchur) [inline scripts/gas-refund-program/staking/common.ts:15]: what's the point of `|| []` here?

Comment (alexshchur) [inline scripts/gas-refund-program/staking/common.ts:15]:
```suggestion
  const minStakeHeldDuringVirtualLockup = (differentialStates || []).reduce(
```

wdyt?

Comment (mwamedacen) [reply scripts/gas-refund-program/staking/common.ts:15]: no point, old code forgot to clean

File: scripts/gas-refund-program/staking/common.ts [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+  if (!differentialStates || differentialStates.length === 0)
+    return stakeAtStartOfVirtualLockup;
+
+  differentialStates.sort(timeseriesComparator);
+
+  const minStakeHeldDuringVirtualLockup = differentialStates.reduce(
+    (minStake, diffStateAtT) => {
+      if (
+        diffStateAtT.timestamp < startOfVirtualLockupPeriod ||
+        diffStateAtT.timestamp > timestamp
```
Comment (alexshchur) [inline scripts/gas-refund-program/staking/common.ts:21]: can also early return if `|| stakeAtT.value.gt(0)`

File: scripts/gas-refund-program/staking/common.ts [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+      ) {
+        return minStake;
+      }
+
+      const newStake = minStake.plus(diffStateAtT.value);
+      const _minStake = BigNumber.min(minStake, newStake);
```
[the remark this answers is not in the document]
Comment (mwamedacen) [reply scripts/gas-refund-program/staking/common.ts:27]: you're actually right

File: scripts/gas-refund-program/staking/common.ts (hunk 1 above)
Comment (mwamedacen) [reply scripts/gas-refund-program/staking/common.ts]: b406ca719917aa429ba807cc235138868422d8da

File: scripts/gas-refund-program/staking/common.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (mwamedacen) [reply scripts/gas-refund-program/staking/common.ts]: b406ca719917aa429ba807cc235138868422d8da

**Review by alexshchur — APPROVED** [review]
as said in PM, I can't say that I've comprehended it to the extend I would like to, but I don't see any particular flaws 👍

**Pull request merged**